src/index/VocabularyGenerator.cpp

```diff
@@ -75,86 +70,166 @@ size_t mergeVocabulary(const std::string& basename, size_t numFiles,
     }
   }
 
-  // keep track of the last seen word to correctly handle duplicates
-  std::string lastWritten = "";
-  // the number of words we have written. This also is the global Id of the next
-  // word we see, unless it is is equal to the previous word
-  size_t totalWritten = 0;
-  bool firstLangPredSeen = false;
-  *langPredLowerBound = 0;
-  *langPredUpperBound = 0;
+  std::vector<QueueWord> sortedBuffer;
+  sortedBuffer.reserve(_bufferSize + 2);
+
+  std::future<void> writeFuture;
 
   // start k-way merge
   while (!queue.empty()) {
-    auto top = queue.top();
+    sortedBuffer.push_back(std::move(queue.top()));
     queue.pop();
+    auto i = sortedBuffer.back()._partialFileId;
+
+    if (sortedBuffer.size() >= _bufferSize) {
+      auto task = [this, buf = std::move(sortedBuffer)]() {
+        this->writeQueueWordsToIdVec(buf);
+      };
+      sortedBuffer.clear();
+      sortedBuffer.reserve(_bufferSize + 2);
+      // wait for the last batch
+
+      if (writeFuture.valid()) {
+        LOG(INFO) << "Waiting for the asynchronous write to finish\n";
+        writeFuture.get();
+      }
+      writeFuture = std::async(task);
+      // we have moved away our buffer, start over
+    }
+
+    // add next word from the same infile to the priority queue
+    if (endOfFile[i]) {
+      continue;
+    }  // file is exhausted, nothing to add
+
+    endOfFile[i] = true;
+    uint32_t len;
+    if (infiles[i].read((char*)&len, sizeof(len))) {
+      std::string word(len, '\0');
+      infiles[i].read(&(word[0]), len);
+      Id id;
+      infiles[i].read((char*)&id, sizeof(id));
+      queue.push(QueueWord(word, i, id));
+      endOfFile[i] = false;
+    }
+  }
+  if (writeFuture.valid()) {
+    writeFuture.get();
+  }
+  if (!sortedBuffer.empty()) {
+    writeQueueWordsToIdVec(sortedBuffer);
+  }
+  VocMergeRes result;
+  result._numWordsTotal = _totalWritten;
+  result._langPredLowerBound = _langPredLowerBound;
+  result._langPredUpperBound = _langPredUpperBound;
+  return result;
+}
 
-    // avoid duplicates
-    if (top._value != lastWritten) {
-      lastWritten = top._value;
+// ________________________________________________________________________________
+void VocabularyMerger::writeQueueWordsToIdVec(
+    const std::vector<QueueWord>& buffer) {
+  LOG(INFO) << "Start writing a batch of merged words\n";
+
+  auto bufSize = _bufferSize / 5;
+  std::future<void> writeFut;
+  std::vector<std::pair<size_t, std::pair<size_t, size_t>>> writeBuf;
+  writeBuf.reserve(bufSize);
+  // avoid duplicates
+  for (const auto& top : buffer) {
+    if (top._value != _lastWritten) {
+      _lastWritten = top._value;
 
       // write the new word to the vocabulary
       if (top._value < string({EXTERNALIZED_LITERALS_PREFIX})) {
```

Comment:
Is this explicit `string` constructor needed?

Reply:
Since `EXTERNALIZED_LITERALS_PREFIX` is a single `char` and thus a number, we
have to do it like this (I did not invent this but merely checked in other places how this has to be done to make it work).

The other question is whether `EXTERNALIZED_LITERALS_PREFIX` should be completely refactored to a `string` or `const char*`

Reply:
I understand, you decide I don't really have a preference